consensus/e2e_tests/state_sync_test.go

```diff
@@ -179,3 +181,132 @@ func TestStateSync_ServerGetBlock_FailNonExistingBlock(t *testing.T) {
+	leaderId := typesCons.NodeId(3)
+	require.Equal(t, uint64(leaderId), testHeight%uint64(numberOfValidators)) // Uses our deterministic round robin leader election
+	testRound := uint64(6)
+	leader := pocketNodes[leaderId]
+	consensusPK, err := leader.GetBus().GetConsensusModule().GetPrivateKey()
+	require.NoError(t, err)
+
+	// Prepare unsynched node info
+	unsynchedNode := pocketNodes[2]
+	unsynchedNodeId := typesCons.NodeId(2)
+	unsynchedNodeHeight := uint64(2)
+	unsynchedNodeModImpl := GetConsensusModImpl(unsynchedNode)
+
+	// Placeholder block
+	blockHeader := &coreTypes.BlockHeader{
+		Height:            testHeight,
+		StateHash:         stateHash,
+		PrevStateHash:     "",
+		ProposerAddress:   consensusPK.Address(),
+		QuorumCertificate: nil,
+	}
+	block := &coreTypes.Block{
+		BlockHeader:  blockHeader,
+		Transactions: make([][]byte, 0),
+	}
+
+	leaderConsensusModImpl := GetConsensusModImpl(leader)
+	leaderConsensusModImpl.MethodByName("SetBlock").Call([]reflect.Value{reflect.ValueOf(block)})
+
+	// Set the unsynched node to height (1) and round (1)
+	// Set rest of the nodes to the same height (3) and round (6)
+	for id, pocketNode := range pocketNodes {
+		consensusModImpl := GetConsensusModImpl(pocketNode)
+		if id == unsynchedNodeId {
+			consensusModImpl.MethodByName("SetHeight").Call([]reflect.Value{reflect.ValueOf(unsynchedNodeHeight)})
+		} else {
+			consensusModImpl.MethodByName("SetHeight").Call([]reflect.Value{reflect.ValueOf(testHeight)})
+		}
+		consensusModImpl.MethodByName("SetStep").Call([]reflect.Value{reflect.ValueOf(testStep)})
+		consensusModImpl.MethodByName("SetRound").Call([]reflect.Value{reflect.ValueOf(testRound)})
+		utilityContext, err := pocketNode.GetBus().GetUtilityModule().NewContext(int64(testHeight))
+		require.NoError(t, err)
+		consensusModImpl.MethodByName("SetUtilityContext").Call([]reflect.Value{reflect.ValueOf(utilityContext)})
+	}
+
+	// Debug message to start consensus by triggering first view change
+	for _, pocketNode := range pocketNodes {
+		TriggerNextView(t, pocketNode)
+	}
+	advanceTime(t, clockMock, 10*time.Millisecond)
+
+	// Assert that unsynched node has a separate view of the network than the rest of the nodes
+	newRoundMessages, err := WaitForNetworkConsensusEvents(t, clockMock, eventsChannel, consensus.NewRound, consensus.Propose, numberOfValidators*numberOfValidators, 250, true)
+	require.NoError(t, err)
+	for nodeId, pocketNode := range pocketNodes {
+		nodeState := GetConsensusNodeState(pocketNode)
+		if nodeId == unsynchedNodeId {
+			assertNodeConsensusView(t, nodeId,
+				typesCons.ConsensusNodeState{
+					Height: unsynchedNodeHeight,
+					Step:   testStep,
+					Round:  uint8(testRound + 1),
+				},
+				nodeState)
+		} else {
+			assertNodeConsensusView(t, nodeId,
+				typesCons.ConsensusNodeState{
+					Height: testHeight,
+					Step:   testStep,
+					Round:  uint8(testRound + 1),
+				},
+				nodeState)
+		}
+		require.Equal(t, false, nodeState.IsLeader)
```

Comment:
No, because currently the check is performed on the `NewRound` messages received from all peers. Nodes will set the leader once we broadcast NewRound responses to all peers, in the following lines. I added a check to the next `assertNodeConsensusView()` for the `leaderId`.